### PR TITLE
Manage npm dependencies

### PR DESCRIPTION
**Pull request by avan1235**
Migrate all placeholder projects to kotlin multiplatform (nothing changed from the top view, still calling `fullDistBuild` is enough to build).

Added sample of using npm dependency as a part of pure sample.

### CODE REVIEW

**Comment by jonnyzzz** [comment]
Thanks for the PR!

File: placeholder-projects/buildSrc/src/main/kotlin/Build.kt [hunk 1]
```diff
@@ -1,58 +1,68 @@
...
+  private const val OUTPUT_DIR_ENV = "OUTPUT_DIR"
+
+  private const val WS = "[ \\t]*"
+
   private val REACT_REGEX = noArgsFunctionCallRegex("useReact")
 
   private val COMPOSE_REGEX = noArgsFunctionCallRegex("useCompose")
 
-  private val MAIN_REGEX = Regex("fun[ \\t]main[ \\t]*[(][ \\t]*[)]")
+  private val MAIN_REGEX = Regex("fun${WS}main${WS}[(]${WS}[)]")
```
Comment (jonnyzzz) [inline placeholder-projects/buildSrc/src/main/kotlin/Build.kt]: Now it'd accept `funmain()` too. A test could be really helpful

File: placeholder-projects/compose/build.gradle.kts [hunk 2]
```diff
@@ -17,12 +18,14 @@ kotlin {
...
-      implementation(project(":pure"))
-    }
+  val jsMain by sourceSets.getting
+
+  val addNpmDependencies = project.addSubprojectsTasks(jsMain, "jsBrowserProductionWebpack")
+
+  jsMain.dependencies {
+    implementation(project(":pure"))
+    implementation(compose.web.core)
+    implementation(compose.runtime)
```
Comment (jonnyzzz) [inline placeholder-projects/compose/build.gradle.kts:30]: Kotlin serialization dependency is missing

File: placeholder-projects/pure/src/main/kotlin/SerializationTrick.kt [hunk 3]
```diff
@@ -1,21 +0,0 @@
-import kotlinx.browser.window
-import kotlinx.serialization.decodeFromString
-import kotlinx.serialization.json.Json
-import kotlin.js.json
-
-inline fun <reified T> fetchJson(url: String, crossinline data: (T) -> dynamic /*Unit or Promise*/) {
```
Comment (jonnyzzz) [inline placeholder-projects/pure/src/main/kotlin/SerializationTrick.kt:6]: could we keep that, but have it fixed?

File: placeholder-projects/react/build.gradle.kts [hunk 4]
```diff
@@ -1,30 +1,34 @@
...
-  implementation("org.jetbrains.kotlin-wrappers:kotlin-react-router-dom:5.2.0-pre.206-kotlin-1.5.10")
-  implementation("org.jetbrains.kotlin-wrappers:kotlin-redux:4.0.5-pre.206-kotlin-1.5.10")
-  implementation("org.jetbrains.kotlin-wrappers:kotlin-react-redux:7.2.3-pre.206-kotlin-1.5.10")
+  val jsMain by sourceSets.getting
+
+  val addNpmDependencies = project.addSubprojectsTasks(jsMain, "jsBrowserProductionWebpack")
+
+  jsMain.dependencies {
+    implementation(project(":pure"))
+    implementation("org.jetbrains.kotlin-wrappers:kotlin-react:17.0.2-pre.206-kotlin-1.5.10")
```
Comment (jonnyzzz) [inline placeholder-projects/react/build.gradle.kts]: we may use BOM for react, as suggested in the recent version of kotlin wrappers

File: placeholder-projects/buildSrc/src/main/kotlin/Build.kt [hunk 5]
```diff
@@ -90,32 +102,35 @@ object Build {
...
+  private fun Project.manageInputFile(inputScript: String, sourceSet: String) {
     val placeholderFile = layout.projectDirectory.file("src/$sourceSet/kotlin/Script.kt").asFile
     val resultScript = modifyInputFile(inputScript)
     placeholderFile.writeText(resultScript)
   }
 
   private fun getOutputDir(): String {
-    val let = getEnv(envName = "OUTPUT_DIR", defaultValue = "./")?.let(::File)?.absoluteFile
-      ?: throw IllegalStateException("OUTPUT_DIR not defined")
+    val let = getEnv(envName = OUTPUT_DIR_ENV, defaultValue = "./")?.let(::File)?.absoluteFile
```
Comment (jonnyzzz) [inline placeholder-projects/buildSrc/src/main/kotlin/Build.kt]: that change may break the Go part, do we need to change the variable names?

**Review by jonnyzzz — COMMENTED** [review]
Let's wait until the hackathon review/judging is completed until the merge. I've added several comment to the changes ;)

File: placeholder-projects/buildSrc/src/main/kotlin/Build.kt (hunk 1 above)
Comment (avan1235) [reply placeholder-projects/buildSrc/src/main/kotlin/Build.kt]: fixed, I will try to add tests package then to this project, this sounds like a good idea but I'll have to think how to do this properly

File: placeholder-projects/buildSrc/src/main/kotlin/Build.kt (hunk 5 above)
Comment (avan1235) [reply placeholder-projects/buildSrc/src/main/kotlin/Build.kt]: I only extracted the value of the string to the const val, but the definition of env variable should not change and it should work with old environment

File: placeholder-projects/react/build.gradle.kts (hunk 4 above)
Comment (avan1235) [reply placeholder-projects/react/build.gradle.kts]: it looks pretty cool, will migrate to bom approach definitely

File: placeholder-projects/buildSrc/src/main/kotlin/Build.kt (hunk 1 above)
Comment (avan1235) [reply placeholder-projects/buildSrc/src/main/kotlin/Build.kt]: The tests from buildSrc are run before every compilation of this module and if they fail the build is not continued. I didn't manage to extract them somewhere else but it seems to be a good enough approach (but there is no possibility to run them manually or I didn't find it)

**Comment by avan1235** [comment]
@jonnyzzz I fixed the `fetchJson` and tested locally with people in space sample (the js was loaded manually, not compiled by server). I will fix the sample to use provided method when this PR is deployed after hackathon review.

I did tiny changes to html examples because we didn't check this in firefox and when I did, it doesn't work in expected way when scaling was used

**Comment by avan1235** [comment]
Hi @jonnyzzz 😃  how are the things? Could we merge this and maybe publish to work on some other feature like proper dependencies imports not only for example projects?

**Comment by jonnyzzz** [comment]
Thanks for the PR :)

The results are now open. Our project has not got any prize. We are in the middle of the grades table

**Comment by avan1235** [comment]
What a pity! Btw have you tried deploying this version on server?

**Comment by jonnyzzz** [comment]
No. We may try.

I figured out we may make compilation even faster! But that would requite
SQS and Gradle --continious mode :)

**Comment by avan1235** [comment]
I can take a look on this when working on the cleaning of imports of dependencies if you wish (probably on weekend)

**Comment by jonnyzzz** [comment]
Thanks! I guess we need first to decide how the project would go further. It's not clear to me right now. Do you have a vision for that?

**Pull request merged**